File: implementations/go/examples/internal/domain/card/repository.go
package card

import "context"

type FindQuery struct {
	Page      int
	Take      int
	CardID    string
	OwnerID   string
	AccountID string
	Status    []Status
}

// Query is a Query-only interface that exposes only read-only lookup
// methods. Query Handlers must depend only on this interface so they have
// no access to write methods (SaveCard) (the same CQRS separation idiom as
// account.Query — cqrs-pattern.md).
//
// Lookups are unified into a single FindCards method, following the root's
// find<Noun>s convention — there is no dedicated single-record lookup
// method. Callers use FindOne (a helper provided by this package) to call
// FindCards with Take: 1 and pull out the first result (the same idiom as
// account.FindOne).
type Query interface {
	FindCards(ctx context.Context, q FindQuery) ([]*Card, int, error)
}

// Repository is a Command-only interface that adds a write method (SaveCard) on top of Query's read methods.
type Repository interface {
	Query
	SaveCard(ctx context.Context, card *Card) error
}

// FindOne is a helper that wraps the repeated single-record lookup pattern
// (call FindCards with Take: 1, then pull out the first result, or
// ErrNotFound if there is none) (the same idiom as account.FindOne).
func FindOne(ctx context.Context, q Query, cardID, ownerID string) (*Card, error) {
	cards, _, err := q.FindCards(ctx, FindQuery{CardID: cardID, OwnerID: ownerID, Take: 1})
	if err != nil {
		return nil, err
	}
	if len(cards) == 0 {
		return nil, ErrNotFound
	}
	return cards[0], nil
}

File: implementations/fastapi/examples/src/card/domain/repository.py
from abc import ABC, abstractmethod

from .card import Card


class CardQuery(ABC):
    """A read-only interface — for the Query Handler only. Never exposes a write method
    such as `save_card()` (see cqrs-pattern.md). Shares its contract with `CardRepository`
    (the write model) but is a separate type — a Query Handler must always depend only on
    this type (the same as the account domain's AccountQuery).

    The lookup method is unified into a single `find_cards(...)` (the same convention as the
    account domain's find_accounts and the payment domain's find_payments) — what used to be
    split into find_by_id/find_by_account has been unified into a single method + optional
    filter keyword arguments. A single-item lookup is expressed with `take=1`.
    """

    @abstractmethod
    async def find_cards(
        self,
        page: int,
        take: int,
        card_id: str | None = None,
        owner_id: str | None = None,
        account_id: str | None = None,
        status: list[str] | None = None,
    ) -> tuple[list[Card], int]: ...


class CardRepository(CardQuery, ABC):
    @abstractmethod
    async def save_card(self, card: Card) -> None: ...
